src/main/java/com/tabnine/Initializer.kt

```diff
@@ -0,0 +1,88 @@
+import com.tabnine.lifecycle.BinaryStateService
+import com.tabnine.lifecycle.TabnineUpdater
+import com.tabnine.logging.initTabnineLogger
+import com.tabnine.notifications.ConnectionLostNotificationHandler
+import com.tabnine.userSettings.AppSettingsState.Companion.instance
+import java.util.concurrent.atomic.AtomicBoolean
+
+class Initializer : PreloadingActivity(), StartupActivity {
+    private var binaryNotificationsLifecycle: BinaryNotificationsLifecycle? = null
+    private var binaryPromotionStatusBarLifecycle: BinaryPromotionStatusBarLifecycle? = null
+    override fun preload(indicator: ProgressIndicator) {
+        initialize()
+    }
+
+    override fun runActivity(project: Project) {
+        initialize()
+    }
+
+    private fun initialize() {
+        val shouldInitialize = !(initialized.getAndSet(true) || ApplicationManager.getApplication().isUnitTestMode)
+        if (shouldInitialize) {
+            Logger.getInstance(javaClass)
+                .info(
+                    "Initializing for ${Config.CHANNEL}, plugin id = ${StaticConfig.TABNINE_PLUGIN_ID_RAW}"
+                )
+            connectionLostNotificationHandler.startConnectionLostListener()
+            ServiceManager.getService(BinaryStateService::class.java).startUpdateLoop()
+            initTabnineLogger()
+            if (!Config.IS_ON_PREM) {
+                binaryNotificationsLifecycle = DependencyContainer.instanceOfBinaryNotifications()
+                binaryPromotionStatusBarLifecycle = DependencyContainer.instanceOfBinaryPromotionStatusBar()
+                binaryNotificationsLifecycle?.poll()
+                binaryPromotionStatusBarLifecycle?.poll()
+                CapabilitiesService.getInstance().init()
+                TabnineUpdater.pollUpdates()
+                PluginInstaller.addStateListener(DependencyContainer.instanceOfUninstallListener())
+            } else if (Config.IS_ON_PREM && !dialogShown.getAndSet(true)) {
+                val cloud2Url = StaticConfig.getTabnineEnterpriseHost()
+                if (cloud2Url.isPresent) {
+                    Logger.getInstance(javaClass)
+                        .info(String.format("Tabnine Enterprise host is configured: %s", cloud2Url.get()))
+                    Utils.setCustomRepository(cloud2Url.get())
+                } else {
+                    Logger.getInstance(javaClass)
+                        .warn(
+                            "Tabnine Enterprise host is not configured, showing some nice dialog"
+                        )
+                    ApplicationManager.getApplication().invokeLater(
+                        Runnable {
+                            val dialog = TabnineEnterpriseUrlDialogWrapper(null)
+                            if (dialog.showAndGet()) {
+                                val url = dialog.inputData
+                                instance.cloud2Url = url
+                                Utils.setCustomRepository(url)
```

Comment:
if you're calling `replaceCustomRepository` inside the setter, why do you call `setCustomRepository` here? is it possible that you got misled by the setter thingy...?